### PR TITLE
Replace merge by merge!

### PR DESCRIPTION
**Pull request by glaucocustodio**
As described [here](https://github.com/JuanitoFatas/fast-ruby#hashmerge-vs-hashmerge-code), `merge!` is 24x faster than `merge`.

### CODE REVIEW

**Comment by dblock** [comment]
See build and rubocop errors, there's notably one you can do a lot faster than `merge!`. 

Generally you cannot be modifying the hash being passed in, so not all your changes of `merge` for `merge!` are ok. Finally if you're going to be doing `merge!`, you don't need to assign the result, it's modifying the hash in-place.

**Comment by glaucocustodio** [comment]
Take a look now please @dblock.

File: lib/grape/cookies.rb [hunk 1]
```diff
@@ -32,7 +32,7 @@ def each(&block)
     end
 
     def delete(name, opts = {})
-      options = opts.merge(value: 'deleted', expires: Time.at(0))
+      options = opts.merge!(value: 'deleted', expires: Time.at(0))
```
Comment (dblock) [inline lib/grape/cookies.rb]: This modifies opts being passed in, you cannot do that.

File: lib/grape/dsl/desc.rb [hunk 2]
```diff
@@ -51,7 +51,7 @@ def desc(description, options = {}, &config_block)
           end
           options = config_class.settings
         else
-          options = options.merge(description: description)
+          options[:description] = description
```
Comment (dblock) [inline lib/grape/dsl/desc.rb]: This modifies options passed in, you cannot do that.

**Comment by dblock** [comment]
Most of these continue to have the same problem: modifying the `Hash` argument being passed in. This *will* have unpredictable side effects - It might be faster, but it only works because we get lucky or we have to think hard about where that hash came from.

File: lib/grape/cookies.rb (hunk 1 above)
Comment (glaucocustodio) [inline lib/grape/cookies.rb]: Why? Both `options` and `opts` are local variables, they aren't used somewhere else except there.

File: lib/grape/dsl/desc.rb (hunk 2 above)
Comment (glaucocustodio) [inline lib/grape/dsl/desc.rb]: Same here, no?!

File: lib/grape/cookies.rb (hunk 1 above)
Comment (dblock) [inline lib/grape/cookies.rb]: When you say local, local to the class? Clearly here `opts` is not a local variable. 

```ruby
2.2.4 :001 > def doit(opts)
2.2.4 :002?>     opts[:foobar] = 1
2.2.4 :003?>   end
 => :doit 
2.2.4 :004 > 
2.2.4 :005 >   opts = {}
 => {} 
2.2.4 :006 > doit(opts)
 => 1 
2.2.4 :007 > p opts
{:foobar=>1}
 => {:foobar=>1} 
```

Modifying a hash within the function causes a side effect of modifying the caller's hash. So it's bad practice, finding bugs caused by such side effect is really hard.

Comment (glaucocustodio) [inline lib/grape/cookies.rb]: 😯 I thought the arguments were copied, but they are the same to those which are passed to the method. Very nice explanation 👏.

**Comment by dblock** [comment]
Lets close this, thanks for hanging in here @glaucocustodio.

**Pull request closed without merging**